sql/hive/src/main/scala/org/apache/spark/sql/hive/hiveUDFs.scala

```diff
@@ -194,47 +183,52 @@ private[hive] case class HiveGenericUDF(
 
   override protected def withNewChildrenInternal(newChildren: IndexedSeq[Expression]): Expression =
     copy(children = newChildren)
-  override protected def doGenCode(ctx: CodegenContext, ev: ExprCode): ExprCode = {
-    val refTerm = ctx.addReferenceObj("this", this)
-    val childrenEvals = children.map(_.genCode(ctx))
-
-    val setDeferredObjects = childrenEvals.zipWithIndex.map {
```

Comment:
codegen is performance critical. Previously, we generate code to directly set values for `deferredObjects`, but in this PR we always create a `Object[]` to wrap the arguments, which can be bad for performance.

This is a signal that `Invoke` doesn't work very well for Hive UDF. It's still valuable to have something like `HiveGenericUDFInvokeAdapter` to share code between interpreted code path and codegen code path.

Reply:
@cloud-fan 
Should I continue to rewrite HiveSimpleUDF with Invoke 
Or the pr https://github.com/apache/spark/pull/39865 is ok?

Reply:
I think we should refactor `HiveGenericUDF` first, then follow it to implement codegen of `HiveSimpleUDF`.

What the refactor should do is to add something like `HiveGenericUDFInvokeAdapter` in this PR, to keep all the states. Then `HiveGenericUDF` just manipulates this stateful object in both interpreted code path and codegen.

Reply:
I have submitted a new pr: https://github.com/apache/spark/pull/40394 to refactor `HiveGenericUDF`.

Reply:
Follow it to implement codegen of HiveSimpleUDF: https://github.com/apache/spark/pull/40397